### PR TITLE
Remove completed tasks from pending queue

### PR DESCRIPTION
**Pull request by payes**
Fixes openebs/openebs#1298

### CODE REVIEW

File: pkg/port/iscsit/session.go [hunk 1]
```diff
@@ -286,6 +286,22 @@ func (tq *taskQueue) Pop() interface{} {
 	return item
 }
 
+func (tq *taskQueue) Delete(x interface{}) {
+	item := x.(*iscsiTask)
```
Comment (AmitKumarDas) [inline pkg/port/iscsit/session.go]: Just check the cases where x cannot be type casted to iscsiTask

Comment (AmitKumarDas) [reply pkg/port/iscsit/session.go]: Please put some unit test code to prove the logic.

File: pkg/port/iscsit/session.go [hunk 2]
```diff
@@ -286,6 +286,22 @@ func (tq *taskQueue) Pop() interface{} {
...
+	for i, t := range old {
+		if t == item {
+			if i == 0 && n == 1 {
+				tq = nil
+				break
+			}
+			*tq = append(old[:i], old[i+1:]...)
+			break
+		}
+	}
```
Comment (AmitKumarDas) [inline pkg/port/iscsit/session.go]: Is it all OK if the item is not found and hence not deleted. Any logs to be printed.?

File: pkg/port/iscsit/session.go [hunk 3]
```diff
@@ -286,6 +286,22 @@ func (tq *taskQueue) Pop() interface{} {
...
+	item := x.(*iscsiTask)
+	old := *tq
+	n := len(old)
+	for i, t := range old {
+		if t == item {
+			if i == 0 && n == 1 {
+				tq = nil
+				break
+			}
+			*tq = append(old[:i], old[i+1:]...)
```
Comment (AmitKumarDas) [inline pkg/port/iscsit/session.go]: Does this mean the item memory is free to be garbage collected after this line?

Comment (payes) [reply pkg/port/iscsit/session.go]: yes

File: pkg/port/iscsit/session.go (hunk 2 above)
Comment (payes) [reply pkg/port/iscsit/session.go]: this function is only entered when there is an item to be deleted, added log will be an unreachable code

File: pkg/port/iscsit/session.go (hunk 1 above)
Comment (payes) [reply pkg/port/iscsit/session.go]: We don't have a go test framework in this repository, libiscsi test suite in jiva takes care of the test cases. This has been manually tested by verifying that the memory leak stops after applying this patch.

File: pkg/port/iscsit/session.go (hunk 3 above)
Comment (gila) [reply pkg/port/iscsit/session.go]: We seem to walk this list twice in the same function  -- first to get a reference to the task (iscsid.go:828), then we process it after which we go searching for it again, in effect making it n^2. Perhaps it's out of scope here, an avl tree or something would be better? In any caseaA function that pops the iscsiTask would prevent that and would also prevent the leak that is still there on iscsid.go:847.

For this PR, @payes are you sure we are not leaking on line 847?

**Pull request merged**